File: tests/db/support/postgres.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import type { ResolvedConnectionConfig } from '@shared/protocol/engine-ops';
import { PostgreSqlContainer, type StartedPostgreSqlContainer } from '@testcontainers/postgresql';
import { Client } from 'pg';
import { resolveDockerHost } from './docker';

resolveDockerHost();

const IMAGE = 'postgres:17-alpine';
const PASSWORD = 'kira';
const DATABASE = 'kira_test';
const STARTUP_TIMEOUT_MS = 120_000;
const BIG_ROWS = 1_000_000;
const SEED_SQL_PATH = resolve(__dirname, '../fixtures/0001_seed.sql');

export interface PgFixture {
  container: StartedPostgreSqlContainer;
  config: ResolvedConnectionConfig; // ready to hand to the adapter
  uri: string;
  stop(): Promise<void>;
}

// One container per test process, started lazily on first call and reused by every later call —
// starting a fresh container per test would make the suite unusable (§11b).
let memoized: Promise<PgFixture> | null = null;

export function startPostgres(opts?: { seedBigTable?: boolean }): Promise<PgFixture> {
  if (!memoized) memoized = start(opts);
  return memoized;
}

async function start(opts?: { seedBigTable?: boolean }): Promise<PgFixture> {
  // @testcontainers/postgresql's own wait strategy (Wait.forAll([forHealthCheck(),
  // forListeningPorts()]), with a pg_isready-based healthcheck it installs automatically) solves
  // the classic Postgres double-boot-during-init flake by polling the real server rather than
  // matching a log line twice — no manual Wait.forLogMessage(..., 2) needed here anymore.
  const container = await new PostgreSqlContainer(IMAGE)
    .withDatabase(DATABASE)
    .withUsername('postgres')
    .withPassword(PASSWORD)
    .withStartupTimeout(STARTUP_TIMEOUT_MS)
    .start();

  const host = container.getHost();
  const port = container.getPort();
  const uri = `postgres://postgres:${PASSWORD}@${host}:${port}/${DATABASE}`;

  const seedClient = new Client({ connectionString: uri });
  await seedClient.connect();
  try {
    await seedClient.query(readFileSync(SEED_SQL_PATH, 'utf8'));
    // Gated: P1 only needs to prove the harness can build the table, P2's paging is what
    // actually consumes 1M rows. Default true so the harness is exercised end to end unless a
    // caller opts out for a faster run.
    if (opts?.seedBigTable ?? true) {
      await seedClient.query(
        'INSERT INTO app.big_rows SELECT i, md5(i::text) FROM generate_series(1, $1) i',
        [BIG_ROWS],
      );
      await seedClient.query('ANALYZE app.big_rows');
    }
  } finally {
    await seedClient.end();
  }

  const now = new Date().toISOString();
  const config: ResolvedConnectionConfig = {
    id: 'test-postgres',
    sortOrder: 0,
    createdAt: now,
    updatedAt: now,
    name: 'Test Postgres',
    kind: 'postgres',
    color: 'blue',
    mode: 'fields',
    readOnly: false,
    host,
    port,
    database: DATABASE,
    username: 'postgres',
    uri: null,
    options: {},
    password: PASSWORD,
  };

  return {
    container,
    config,
    uri,
    async stop() {
      // Playwright's workers:1 config runs every UI spec file sequentially in the same worker
      // process, sharing this module's state — without resetting `memoized`, a later spec file's
      // startPostgres() would return this now-dead container instead of starting a fresh one.
      memoized = null;
      await container.stop();
    },
  };
}

const SCROLL_GRID_COLUMNS = 60;
const SCROLL_GRID_ROWS = 5_000;

// P29 D14: neither the standard seed's `big_rows` (many rows, few columns) nor `wide_table` (few
// rows, many columns) can show the scroll-render gap — this creates the wide-AND-tall shape P29's
// own F5/F8 need, against whichever container `startPostgres()` handed back (never
// `tests/db/fixtures/0001_seed.sql`, which every other suite's row counts/ordering already depend
// on staying unchanged). Moved here from `tests/e2e/support/pg.ts` (P57 M5: `tests/e2e/` retired)
// — `scripts/capture-postgres-tree.ts`'s own `seedScrollGrid` step kind still needs it to
// reproduce `tests/ui/budgets.spec.ts`/`perf.spec.ts`'s real `app.scroll_grid` captures.
/** `app.scroll_grid`: 60 text columns x 5 000 rows, one integer primary key, no foreign keys. */
export async function seedScrollFixture(uri: string): Promise<void> {
  const client = new Client({ connectionString: uri });
  await client.connect();
  try {
    const columns = Array.from({ length: SCROLL_GRID_COLUMNS }, (_, i) => `col${i + 1}`);
    const createCols = columns.map((c) => `${c} text NOT NULL`).join(',\n        ');
    await client.query(`
      CREATE TABLE app.scroll_grid (
        id integer PRIMARY KEY,
        ${createCols}
      );
    `);
    const selectCols = columns
      .map((c, i) => `'row ' || i || ' col ${i + 1}' AS ${c}`)
      .join(',\n        ');
    await client.query(`
      INSERT INTO app.scroll_grid
      SELECT i AS id, ${selectCols}
      FROM generate_series(1, ${SCROLL_GRID_ROWS}) i;
    `);
    await client.query('ANALYZE app.scroll_grid');
  } finally {
    await client.end();
  }
}

File: shell/internal/adapters/testsupport/postgres.go
// Package testsupport is the Go analogue of tests/db/support/*.ts (A19): the Docker gate, real
// container startup, and seed-SQL loading each adapter's own acceptance spec needs. One container
// per test binary, started lazily on first call and reused by every later call in the same
// process — a fresh container per test would make the suite unusable (§11b).
package testsupport

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"testing"
	"time"

	"github.com/testcontainers/testcontainers-go"
	tcpostgres "github.com/testcontainers/testcontainers-go/modules/postgres"
	"github.com/testcontainers/testcontainers-go/wait"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// DockerUnavailableMessage mirrors tests/db/support/docker.ts's own DOCKER_UNAVAILABLE_MESSAGE,
// re-pointed at this environment's own startup procedure (AGENTS.md's Docker section) rather than
// Colima's, since this is the Go test tier, not the dev machine's.
const DockerUnavailableMessage = "Docker daemon unreachable — see AGENTS.md's Docker section for " +
	"how to start it (nohup dockerd, or colima start on macOS) and retry."

const dockerProbeTimeout = 5 * time.Second

// IsDockerAvailable is docker.ts's isDockerAvailable.
func IsDockerAvailable() bool {
	ctx, cancel := context.WithTimeout(context.Background(), dockerProbeTimeout)
	defer cancel()
	return exec.CommandContext(ctx, "docker", "info").Run() == nil
}

// PgFixture is postgres.ts's PgFixture.
type PgFixture struct {
	URI       string
	Config    model.ResolvedConnectionConfig
	container testcontainers.Container
}

var pgMemo fixture[PgFixture]

// repoRoot resolves the repository root relative to this source file, so the seed SQL path does
// not depend on the test binary's own working directory.
func repoRoot() string {
	_, thisFile, _, _ := runtime.Caller(0)
	// this file: shell/internal/adapters/testsupport/postgres.go
	return filepath.Join(filepath.Dir(thisFile), "..", "..", "..", "..")
}

const (
	image    = "postgres:17-alpine"
	password = "kira"
	database = "kira_test"
)

// StartPostgres is postgres.ts's startPostgres. Skips the test (never a silent pass, never a hard
// failure that makes go test ./internal/adapters/postgres unusable without Docker) when the
// daemon is unreachable.
//
// Memoized process-wide by fixture[T] (bun:test's own beforeAll/afterAll-per-file precedent,
// §11b/B15) — termination is deliberately NOT wired to any one test's t.Cleanup: see fixture.go's
// own doc comment for why. Call StopPostgres from the package's own TestMain instead.
func StartPostgres(t *testing.T) *PgFixture {
	t.Helper()
	if !IsDockerAvailable() {
		t.Skip(DockerUnavailableMessage)
	}
	fixture, err := pgMemo.get(startPostgres)
	if err != nil {
		t.Fatalf("postgres container: %v", err)
	}
	return fixture
}

// StopPostgres terminates the memoized container, if one was ever started. Call once, from the
// test binary's own TestMain, after m.Run() returns — never from an individual test.
func StopPostgres() {
	pgMemo.stop(func(f *PgFixture) { _ = f.container.Terminate(context.Background()) })
}

func startPostgres() (*PgFixture, error) {
	ctx := context.Background()

	// The same wait strategy P58a's own M0 probe confirmed working in this sandbox: Postgres logs
	// "database system is ready to accept connections" twice during initdb's own double-boot, and
	// waiting for the second occurrence is what @testcontainers/postgresql's own pg_isready-based
	// healthcheck solves on the TypeScript side — testcontainers-go's postgres module has no
	// equivalent built-in healthcheck, so this is spelled out explicitly here instead.
	container, err := tcpostgres.Run(ctx, image,
		tcpostgres.WithDatabase(database),
		tcpostgres.WithUsername("postgres"),
		tcpostgres.WithPassword(password),
		testcontainers.WithWaitStrategy(
			wait.ForLog("database system is ready to accept connections").WithOccurrence(2).WithStartupTimeout(120*time.Second),
		),
	)
	if err != nil {
		return nil, err
	}

	host, err := container.Host(ctx)
	if err != nil {
		return nil, err
	}
	port, err := container.MappedPort(ctx, "5432/tcp")
	if err != nil {
		return nil, err
	}
	uri := "postgres://postgres:" + password + "@" + host + ":" + port.Port() + "/" + database

	seedPath := filepath.Join(repoRoot(), "tests", "db", "fixtures", "0001_seed.sql")
	seedSQL, err := os.ReadFile(seedPath)
	if err != nil {
		return nil, err
	}
	if err := seedDatabase(ctx, uri, string(seedSQL)); err != nil {
		return nil, err
	}
	// postgres.ts's own start(): app.big_rows is created by the seed file but populated and
	// ANALYZEd separately (opts?.seedBigTable ?? true) — done unconditionally here since every one
	// of this package's own tests wants the real 1,000,000-row table, not the P1-only empty-table
	// shortcut the TS harness offers callers that don't need paging.
	if err := seedBigRows(ctx, uri); err != nil {
		return nil, err
	}

	now := time.Now().UTC().Format(time.RFC3339Nano)
	hostCopy, portInt := host, int(port.Num())
	cfg := model.ResolvedConnectionConfig{
		ID: "test-postgres", SortOrder: 0, CreatedAt: now, UpdatedAt: now,
		Name: "Test Postgres", Kind: "postgres", Color: "blue", Mode: "fields", ReadOnly: false,
		Host: &hostCopy, Port: &portInt, Database: Strp(database), Username: Strp("postgres"),
		Options: map[string]any{}, Password: Strp(password),
	}
	return &PgFixture{URI: uri, Config: cfg, container: container}, nil
}
